DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,234 @@
+      TestBasicOpsAcInt(q, a, b, c, d, e);
+
+      int _c, _d, _e;
+      TestBasicOpsInt(q, a, b, _c, _d, _e);
+
+      if (c != _c || d != _d || e != _e) {
+        std::cout << "Result mismatch!\n"
+                  << "Kernel BasicOpsInt:   addition = " << _c
+                  << ", multiplication = " << _d << ", division = " << _e
+                  << "\n"
+                  << "Kernel BasicOpsAcInt: addition = " << c
+                  << ", multiplication = " << d << ", division = " << e
+                  << "\n\n";
+        passed = false;
+      }
+    }
+
+    // Kernel `ShiftOp` contains an `ac_int` left shifter and the data type of
+    // the shift amount is a large width signed `ac_int`. On contrast, kernel
+    // `EfficientShiftOp` also contains an `ac_int` left shifter but the data
+    // type of the shift amount is a reduced width unsigned `ac_int`. By
+    // comparing these two kernels, you will find shift operations of `ac_int`
+    // can generate more efficient hardware if the amount to shift by is stored
+    // in a minimally sized unsigned `ac_int`.
+    {
+      my_int14 a = t1, b = t2;
+      my_uint2 _b = t2;
+      my_int14 c, _c;
+      TestShiftOp(q, a, b, c);
+      TestEfficientShiftOp(q, a, _b, _c);
+
+      if (c != _c) {
+        std::cout << "Result mismatch!\n"
+                  << "Kernel ShiftOp: result = " << c << "\n"
+                  << "Kernel EfficientShiftOp: result = " << _c << "\n\n";
+        passed = false;
+      }
+    }
+
+    // Kernel `BitOps` demonstrates bit operations with bit select operator `[]`
+    // and bit slice write operation `set_slc`. Note: An `ac_int` must be
+    // initialized before being access by bit select operator `[]` and bit slice
+    // operations `slc` and `set_slc`, otherwise it is undefined behavior and
+    // will give you unexpected results.
+    {
+      my_int14 a = t1;
+      my_int14 result = TestBitOps(q, a);
+
+      int golden = 0b001111101111;
+
+      if (result != golden) {
+        std::cout << "Kernel BitOps result mismatch!\n"
+                  << "result = 0b" << std::bitset<14>(result) << "\n"
+                  << "golden = 0b" << std::bitset<14>(golden) << "\n\n";
+        passed = false;
+      }
+    }
+  } catch (exception const &e) {
+    // Catches exceptions in the host code.
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.code().value() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+
+  if (passed) {
+    std::cout << "PASSED: all kernel results are correct.\n";
```

Comment:
Meh. Why not just "PASSED" ?

Reply:
Does this really provide more detail? Why not make the FAILED case "FAILED: 1 or more kernels failed"? Or better yet, "FAILED: <n> kernels failed" where you print how many kernels faild?

Reply:
I prefer "PASSED: all kernel results are correct.\n" because I value the consistency among all code samples. 
If it turns out that the community doesn't like this version of print message, all code samples should change it together.

Reply:
Not all code samples do this. Some do, some don't. I also like consistency, but right now we don't have it.

Reply:
> Does this really provide more detail? 
A tiny bit.  It tells you that 'passed' means the results are correct, as opposed to passing some performance test.

This is not worth arguing about, Shuo your call.

Reply:
I disagree. If we plan to commonize this across tutorials, we should pick one. And I think we should commonize across tutorials. I can imagine uses for always printing the same "last line" from the tutorials.